x/mongo/driver/topology/fsm.go

```diff
@@ -228,30 +228,75 @@ func (f *fsm) checkIfHasPrimary() {
 	}
 }
 
-func (f *fsm) updateRSFromPrimary(s description.Server) {
-	if f.SetName == "" {
-		f.SetName = s.SetName
-	} else if f.SetName != s.SetName {
-		f.removeServerByAddr(s.Addr)
-		f.checkIfHasPrimary()
-		return
+// hasStalePrimary returns true if the topology has a primary that is stale.
+func hasStalePrimary(fsm fsm, srv description.Server) bool {
+	if srv.WireVersion == nil {
+		return true
+	}
+
+	// Compare the election ID values of the server and the topology lexicographically.
+	compRes := bytes.Compare(srv.ElectionID[:], fsm.maxElectionID[:])
+
+	if srv.WireVersion.Max < 17 {
+		return compRes == -1 || fsm.maxSetVersion > srv.SetVersion
+	}
+
+	return compRes != 1 && (compRes != 0 || srv.SetVersion < fsm.maxSetVersion)
+}
+
+// transferEVTuple will transfer the ("ElectionID", "SetVersion") tuple from the description server to the topology.
+// If the primary is stale, the tuple will not be transferred, the topology will update it's "Kind" value, and this
+// routine will return "false".
+func transferEVTuple(srv description.Server, fsm *fsm) bool {
+	stalePrimary := hasStalePrimary(*fsm, srv)
+
+	if wireVersion := srv.WireVersion; wireVersion != nil && wireVersion.Max >= 17 {
```

Comment:
Confused about the logic here. I thought the wire version checking was all done in `hasStalePrimary`? The only difference I see pre-6.0 and post-6.0 in `transferEVTuple` is whether we run `replaceServer`. Reading the comment above the function, I thought the logic would read more like:
```
func transferEVTuple(srv description.Server, fsm *fsm) bool {
      stalePrimary := hasStalePrimary(*fsm, srv)
      if stalePrimary {
           fsm.checkIfHasPrimary()
           return false
      }
      fsm.maxElectionID = srv.ElectionID
      fsm.maxSetVersion = srv.SetVersion
      return true
}
```

Reply:
We still have to maintain the [old comparison rules](https://github.com/mongodb/specifications/commit/37c32cf5200cd8da6652c75ac65279c43628fe7a#diff-52a48d3b84e9e90e30fb87b3799c6302750f1ed2817b2abbead44e7a7bb6bb47R1130) which have a different set of logic for a stale primary as well as how to set data on `fsm`.

But I do think this part of the check is redundant somewhere: `wireVersion != nil`.

Reply:
Ah that sounds find to me. 

> But I do think this part of the check is redundant somewhere: wireVersion != nil.

I actually think it's fine as-is. We do also check that wireVersion != nil in `hasStalePrimary` but it sounds like we need to know in both `transferEVTuple` and `hasStalePrimary`.